htlcswitch/hop/iterator.go

```diff
@@ -113,6 +114,108 @@ func (r *sphinxHopIterator) ExtractErrorEncrypter(
 	return extracter(r.ogPacket.EphemeralKey)
 }
 
+// BlindingProcessor is an interface that provides the cryptographic operations
+// required for processing blinded hops.
+type BlindingProcessor interface {
+	// DecryptBlindedHopData decrypts a blinded blob of data using the
+	// ephemeral key provided.
+	DecryptBlindedHopData(ephemPub *btcec.PublicKey,
+		encryptedData []byte) ([]byte, error)
+}
+
+// BlindingKit contains the components required to extract forwarding
+// information for hops in a blinded route.
+type BlindingKit struct {
+	// BlindingPoint holds a blinding point that was passed to the node via
+	// update_add_htlc's TLVs.
+	BlindingPoint *btcec.PublicKey
+
+	// ForwardingInfo uses the ephemeral blinding key provided to decrypt
+	// a blob of encrypted data provided in the onion and obtain the
+	// forwarding information for the blinded hop.
+	ForwardingInfo func(*btcec.PublicKey, []byte) (*ForwardingInfo,
+		error)
+}
+
+// MakeBlindingKit produces a kit that is used to decrypte and decode
+// forwarding information for hops in blinded routes.
+func MakeBlindingKit(processor BlindingProcessor,
+	blindingPoint *btcec.PublicKey, incomingAmount lnwire.MilliSatoshi,
+	incomingCltv uint32) *BlindingKit {
+
+	return &BlindingKit{
+		BlindingPoint: blindingPoint,
+		ForwardingInfo: func(blinding *btcec.PublicKey,
+			data []byte) (*ForwardingInfo, error) {
+
+			decrypted, err := processor.DecryptBlindedHopData(
+				blinding, data,
+			)
+			if err != nil {
+				return nil, fmt.Errorf("decrypt blinded "+
+					"data: %w", err)
+			}
+
+			b := bytes.NewBuffer(decrypted)
+			routeData, err := record.DecodeBlindedRouteData(b)
+			if err != nil {
+				return nil, fmt.Errorf("decode route data: "+
+					"%w", err)
+			}
+
+			if err := ValidateBlindedRouteData(
+				routeData, incomingAmount, incomingCltv,
+			); err != nil {
+				return nil, err
+			}
+
+			return deriveForwardingInfo(
+				routeData, incomingAmount, incomingCltv,
+			)
+		},
+	}
+}
+
+// deriveForwardingInfo calculates forwarding information from the (valid)
+// blinded route data provided and the incoming htlc's information.
+func deriveForwardingInfo(data *record.BlindedRouteData,
+	incomingAmt lnwire.MilliSatoshi, incomingCltv uint32) (*ForwardingInfo,
+	error) {
+
+	// If we have our short channel ID or expiry present, set values in
+	// our forwarding information. We start with the incoming values as
+	// defaults so that they will have the correct values for the final
+	// hop in the blinded route (which does not have relay info set).
+	var (
+		nextHop = Exit
+		expiry  = incomingCltv
+		fwdAmt  = incomingAmt
+	)
+
+	if data.ShortChannelID != nil {
+		nextHop = *data.ShortChannelID
+	}
```

Comment:
do we not need to account for the case where they set the NextNodeID instead of SCID? 

I know that the `next_node_id` will mostly be used for onion messaging and SCID is much preferred for encrypted data since it is more compact. but I dont think the spec explicitly disallows a receiver from using next_node_id instead of scid here.

Reply:
hopefully i've missed or misunderstood something since I know that allowing for both here will make things a bit more complicated.... unless we just use a call back here to query for any SCID that we have to the given node that is big enough for the htlc size?

Reply:
Sigh, dual purpose `encrypted_data` sucks. 

Spec says that `encrypted_data` creators [must](https://github.com/lightning/bolts/blob/master/04-onion-routing.md#requirements) include `short_channel_id` for non-final nodes, so I _think_ that we could just update `ValidateBlindedRouteData` to enforce the presence of a short channel ID? 

That may put us at risk of being too restrictive (but anyone not supplying a short channel ID is not compliant anyway), so possibly something we could leave for a follow up, WDYT? Shouldn't' be too bad to thread the callback through in the `BlindingKit`, but would be nice to try keep the scope down a little :')